### PR TITLE
*: use DDL subscriber updating stats meta

### PR DESCRIPTION
**Pull request by Rustin170506**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/57871

Problem Summary:

### What changed and how does it work?

In this PR, we began using the DDL notifier to handle all DDL events and update the stats meta accordingly.

To maintain the old behavior, we do not register it when the stats lease is less than 1. This means that we won't handle the DDL event for unit tests at all unless you explicitly handle it manually.

In the future, we might want to handle this in unit tests by default. However, doing so could disrupt many existing test cases. For now, I prefer to maintain the old behavior. I plan to migrate to the new approach eventually, but I currently don't have enough time to update all the tests that would break.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] [Manual test](https://github.com/pingcap/tidb/pull/57872#issuecomment-2535350098)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by winoros — APPROVED** [review]

**Comment by Rustin170506** [comment]
There is a data race here. I will fix it later.

File: pkg/ddl/ddl.go [hunk 1]
```diff
@@ -588,18 +588,23 @@ func asyncNotifyEvent(jobCtx *jobContext, e *notifier.SchemaChangeEvent, job *mo
...
-			default:
-				time.Sleep(time.Microsecond * 10)
+	// In test environments, we use a channel-based approach to handle DDL events.
+	// This maintains compatibility with existing test cases that expect events to be delivered through channels.
+	// In production, DDL events are handled by the notifier system instead.
+	if intest.InTest {
+		ch := jobCtx.oldDDLCtx.ddlEventCh
+		if ch != nil {
+		forLoop:
+			for i := 0; i < 10; i++ {
```
Comment (lance6716) [inline pkg/ddl/ddl.go:598]: if it's only a test, we don't need to retry? So the duration of unit test can be reduced

**Review by lance6716 — COMMENTED** [review]
33 / 57 files viewed

File: pkg/statistics/handle/ddl/ddl.go [hunk 2]
```diff
@@ -46,158 +42,18 @@ func NewDDLHandler(
 		ddlEventCh:   make(chan *notifier.SchemaChangeEvent, 1000),
 		statsWriter:  statsWriter,
 		statsHandler: statsHandler,
+		sub:          NewSubscriber(statsHandler),
```
Comment (lance6716) [inline pkg/statistics/handle/ddl/ddl.go:45]:
```suggestion
		sub:          newSubscriber(statsHandler),
```

**Review by lance6716 — APPROVED** [review]
/hold

free to unhold

**Pull request merged**